hnn_core/gui/gui.py

```diff
@@ -628,16 +803,73 @@ def _clear_params(b):
         self.run_button.on_click(_run_button_clicked)
         self.clear_button.on_click(_clear_params)
 
-        # widgets whose changes should trigger viz changes.
-        _vis_related_widgets = [
-            self.widget_viz_layout_selection,
-            self.widget_max_spectral_frequency,
-            self.widget_dipole_scaling,
-            self.widget_dipole_smooth,
-            self.widget_spectrogram_cm,
-        ]
-        for _widget in _vis_related_widgets:
-            _widget.observe(_handle_viz_layout_change, 'value')
+        def _handle_viz_config_layout_change(b):
+            new_key = b['new']
+            with self._log_out:
+                assert new_key in self.analysis_config['gui_config'][
+                    'fig_templates'].keys()
+                assert self.fig_titles.selected_index == len(
+                    self.data['figure_trees'])
+            _init_plot_configs(self._plot_configs,
+                               self.simulation_data,
+                               self.vis_plot_tree,
+                               self.analysis_config,
+                               use_template=True,
+                               render=True)
+
+        self.default_viz_config_selection.observe(
+            _handle_viz_config_layout_change,
+            "value"
+        )
+
+        def _handle_fig_change(b):
+            fig_idx = b['new']
+            n_figs = len(self.figure_trees)
+            if fig_idx == n_figs:
+                self.default_viz_config_selection.disabled = False
+                self.run_plotting_button.description = add_fig_description
+                # selected current
+                self.run_plotting_button.description = add_fig_description
+                _init_plot_configs(self._plot_configs, self.simulation_data,
+                                   self.vis_plot_tree, self.analysis_config,
+                                   render=True)
+            else:
+                self.default_viz_config_selection.disabled = True
+                fig_idx = b['new']
+                if fig_idx is not None:
+                    assert 0 <= fig_idx < len(self.figure_trees)
+                    self.run_plotting_button.description = f"Save \
+                        Fig.{fig_idx+1}"
+                    _init_plot_configs(self._plot_configs,
+                                       self.simulation_data,
+                                       self.figure_trees[fig_idx],
+                                       self.analysis_config, render=True)
+
+        self.fig_titles.observe(_handle_fig_change, names='selected_index')
+
+        def _on_plot_clicked(_):
```

Comment:
I don't like this pattern ... reading the code it's hard to figure out what the function expects as arguments. Just list out the arguments.